File: scripts/import_molecules_to_db.py
"""
Import molecules from public/data/chembl-molecules.json into PostgreSQL.
Stores only: chembl_id, name, smiles, synonyms.

Usage:
  python scripts/import_molecules_to_db.py
"""

import json
import os
from pathlib import Path

import psycopg2
from dotenv import load_dotenv


ROOT = Path(__file__).resolve().parents[1]
ENV_PATH = ROOT / ".env"
if not ENV_PATH.exists():
    ENV_PATH = ROOT / ".env.local"
load_dotenv(dotenv_path=ENV_PATH)

DATA_PATH = ROOT / "public" / "data" / "chembl-molecules.json"

DB_CONFIG = {
    "dbname": os.getenv("DB_NAME", "filspresso"),
    "user": os.getenv("DB_USER", "filspresso_user"),
    "password": os.getenv("DB_PASSWORD"),
    "host": os.getenv("DB_HOST", "localhost"),
    "port": os.getenv("DB_PORT", "5432"),
}


def normalize_synonyms(value):
    if not value:
        return []
    if isinstance(value, list):
        return [str(v).strip() for v in value if str(v).strip()]
    if isinstance(value, str):
        return [value.strip()] if value.strip() else []
    return []


def ensure_schema(conn):
    with conn.cursor() as cur:
        cur.execute(
            """
            CREATE TABLE IF NOT EXISTS molecules (
                id SERIAL PRIMARY KEY,
                chembl_id VARCHAR(64) UNIQUE,
                name VARCHAR(255),
                smiles TEXT NOT NULL,
                synonyms JSONB DEFAULT '[]'::jsonb,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );
            """
        )
        cur.execute("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS chembl_id VARCHAR(64)")
        cur.execute("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS name VARCHAR(255)")
        cur.execute("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS smiles TEXT")
        cur.execute("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS synonyms JSONB DEFAULT '[]'::jsonb")
        cur.execute("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
        cur.execute("CREATE INDEX IF NOT EXISTS idx_molecules_chembl_id ON molecules(chembl_id);")
        cur.execute("CREATE INDEX IF NOT EXISTS idx_molecules_name ON molecules(name);")


def main():
    if not DATA_PATH.exists():
        raise FileNotFoundError(f"Dataset not found: {DATA_PATH}")

    with DATA_PATH.open("r", encoding="utf-8") as f:
        payload = json.load(f)

    molecules = payload.get("molecules", []) if isinstance(payload, dict) else payload
    if not isinstance(molecules, list):
        raise ValueError("Invalid dataset format: expected list or { molecules: [] }")

    conn = psycopg2.connect(**DB_CONFIG)
    try:
        ensure_schema(conn)
        inserted = 0
        updated = 0

        with conn.cursor() as cur:
            for mol in molecules:
                chembl_id = (mol.get("chembl_id") or "").strip() or None
                name = (mol.get("name") or "").strip() or None
                smiles = (mol.get("smiles") or "").strip()
                synonyms = normalize_synonyms(mol.get("synonyms"))

                if not smiles:
                    continue

                cur.execute(
                    """
                    INSERT INTO molecules (chembl_id, name, smiles, synonyms)
                    VALUES (%s, %s, %s, %s::jsonb)
                    ON CONFLICT (chembl_id) DO UPDATE
                    SET name = EXCLUDED.name,
                        smiles = EXCLUDED.smiles,
                        synonyms = EXCLUDED.synonyms
                    """,
                    (chembl_id, name, smiles, json.dumps(synonyms)),
                )
                if cur.rowcount == 1:
                    inserted += 1
                else:
                    updated += 1

        conn.commit()
        print(f"Imported molecules complete. inserted_or_updated={inserted + updated}, inserted={inserted}, updated={updated}")
    finally:
        conn.close()


if __name__ == "__main__":
    main()

File: express-api/scripts/import_molecules_to_db.js
const fs = require("fs");
const path = require("path");
const dotenv = require("dotenv");
const { Pool } = require("pg");

const repoRoot = path.resolve(__dirname, "..", "..");
const envLocalPath = path.join(repoRoot, ".env.local");
const envPath = path.join(repoRoot, ".env");

if (fs.existsSync(envLocalPath)) {
	dotenv.config({ path: envLocalPath });
} else if (fs.existsSync(envPath)) {
	dotenv.config({ path: envPath });
} else {
	dotenv.config();
}

const baseDbConfig = {
	port: parseInt(process.env.DB_PORT || "5432", 10),
	database: process.env.DB_NAME || "filspresso",
	user: process.env.DB_USER || "filspresso_user",
	password: process.env.DB_PASSWORD,
};

function buildHostCandidates() {
	const configured = String(process.env.DB_HOST || "localhost").trim();
	if (!configured || configured === "localhost") {
		return ["localhost"];
	}
	return [configured, "localhost"];
}

async function connectWithFallback() {
	const hosts = buildHostCandidates();
	let lastError;

	for (const host of hosts) {
		const pool = new Pool({ ...baseDbConfig, host });
		try {
			const client = await pool.connect();
			await client.query("SELECT 1");
			console.log(`Connected to PostgreSQL host: ${host}`);
			return { pool, client };
		} catch (error) {
			lastError = error;
			await pool.end().catch(() => undefined);
		}
	}

	throw lastError;
}

function normalizeSynonyms(value) {
	if (!value) return [];
	if (Array.isArray(value)) {
		return value.map((v) => String(v).trim()).filter(Boolean);
	}
	if (typeof value === "string") {
		const trimmed = value.trim();
		return trimmed ? [trimmed] : [];
	}
	return [];
}

async function ensureSchema(client) {
	await client.query(`
		CREATE TABLE IF NOT EXISTS molecules (
			id SERIAL PRIMARY KEY,
			chembl_id VARCHAR(64) UNIQUE,
			name VARCHAR(255),
			smiles TEXT NOT NULL,
			synonyms JSONB DEFAULT '[]'::jsonb,
			created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
		)
	`);

	await client.query("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS chembl_id VARCHAR(64)");
	await client.query("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS name VARCHAR(255)");
	await client.query("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS smiles TEXT");
	await client.query("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS synonyms JSONB DEFAULT '[]'::jsonb");
	await client.query("ALTER TABLE molecules ADD COLUMN IF NOT EXISTS created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP");

	await client.query("CREATE UNIQUE INDEX IF NOT EXISTS idx_molecules_chembl_id_unique ON molecules(chembl_id)");
	await client.query("CREATE INDEX IF NOT EXISTS idx_molecules_name ON molecules(name)");
}

async function upsertByChembl(client, molecule) {
	await client.query(
		`
		INSERT INTO molecules (chembl_id, name, smiles, synonyms)
		VALUES ($1, $2, $3, $4::jsonb)
		ON CONFLICT (chembl_id) DO UPDATE
		SET name = EXCLUDED.name,
			smiles = EXCLUDED.smiles,
			synonyms = EXCLUDED.synonyms
		`,
		[molecule.chembl_id, molecule.name, molecule.smiles, JSON.stringify(molecule.synonyms)],
	);
}

async function upsertBySmilesFallback(client, molecule) {
	const existing = await client.query("SELECT id FROM molecules WHERE smiles = $1 LIMIT 1", [molecule.smiles]);
	if (existing.rows.length > 0) {
		await client.query(
			"UPDATE molecules SET name = COALESCE($1, name), synonyms = $2::jsonb WHERE id = $3",
			[molecule.name, JSON.stringify(molecule.synonyms), existing.rows[0].id],
		);
		return;
	}

	await client.query(
		"INSERT INTO molecules (chembl_id, name, smiles, synonyms) VALUES ($1, $2, $3, $4::jsonb)",
		[molecule.chembl_id, molecule.name, molecule.smiles, JSON.stringify(molecule.synonyms)],
	);
}

async function run() {
	const dataPath = path.join(repoRoot, "public", "data", "chembl-molecules.json");
	if (!fs.existsSync(dataPath)) {
		throw new Error(`Dataset file not found: ${dataPath}`);
	}

	const raw = fs.readFileSync(dataPath, "utf8");
	const parsed = JSON.parse(raw);
	const source = Array.isArray(parsed) ? parsed : parsed.molecules;

	if (!Array.isArray(source)) {
		throw new Error("Invalid dataset format: expected array or { molecules: [] }");
	}

	if (!process.env.DB_PASSWORD) {
		throw new Error("DB_PASSWORD environment variable is required");
	}

	let client;
	let activePool;
	let imported = 0;
	let skipped = 0;

	try {
		const connection = await connectWithFallback();
		activePool = connection.pool;
		client = connection.client;
		await client.query("BEGIN");
		await ensureSchema(client);

		for (const row of source) {
			const smiles = String(row.smiles || "").trim();
			if (!smiles) {
				skipped += 1;
				continue;
			}

			const molecule = {
				chembl_id: String(row.chembl_id || "").trim() || null,
				name: String(row.name || "").trim() || null,
				smiles,
				synonyms: normalizeSynonyms(row.synonyms),
			};

			if (molecule.chembl_id) {
				await upsertByChembl(client, molecule);
			} else {
				await upsertBySmilesFallback(client, molecule);
			}
			imported += 1;
		}

		await client.query("COMMIT");
		console.log(`Molecule import completed. Imported/updated: ${imported}, skipped: ${skipped}`);
	} catch (error) {
		if (client) {
			await client.query("ROLLBACK");
		}
		throw error;
	} finally {
		if (client) client.release();
		if (activePool) {
			await activePool.end();
		}
	}
}

run().catch((error) => {
	console.error("Molecule import failed:", error.message);
	process.exit(1);
});
